---
title: "Rice-PCAs-SNPs"
author: "Aidan Angus-Henry"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
```

### Import the Data

```{r}
Sys.setenv(VROOM_CONNECTION_SIZE="500000") # needed because the lines in this file are very long.
data.geno <- read_csv("../input/Rice_44K_genotypes.csv.gz",
                      na=c("NA","00")) #missing data is denoted as "NA" or "00"
```

```{r}
data.geno <- data.geno %>% select(-`6_17160794...22253`) # remove a duplicate column
```

```{r}
head(data.geno)
```

```{r}
data.geno <- data.geno %>% rename(ID=...1) # Rename the first column to ID
```

### Construct PCA Plot

_Due to the size of the dataset, a randomly selected subset of 10k SNPs will be used_

```{r}
set.seed(0421)
data <- data.geno[,sample(2:ncol(data.geno), 10000)] # Subset the data
id <- select(data.geno, ID) # Extract the ID column
data.geno.10k <- data.frame(id, data) # Combine the ID column and 10k SNP subset
dim(data.geno.10k)
```

```{r}
#convert the data matrix to numbers
geno.numeric <- data.geno.10k[,-1] %>% # -1 to remove the first column, with names.
  lapply(factor) %>% # convert characters to "factors", where each category is internally represented as a number
  as.data.frame() %>% # reformat
  data.matrix() #  convert to numeric

head(geno.numeric[,1:10])
```

```{r}
# Fill in missing data with average genotype
geno.numeric.fill <-
  apply(geno.numeric, 2, function(x) {
    x[is.na(x)] <- mean(x, na.rm=T)
    x})
```

```{r}
geno.pca <- prcomp(geno.numeric.fill, 
            rank.=10) # Compute the PCs, return the first 10

str(geno.pca)  
```

```{r}
head(geno.pca$x)[,1:5] # View the first 5 PCs
```

Determine proportion of variance explained by each PC:
```{r}
pcvar <- geno.pca$sdev^2 # square std dev to get variance
pcvar.pct <- tibble(pctvar=pcvar/sum(pcvar) * 100,
                    PC=1:length(pcvar))

head(pcvar.pct)
```

```{r}
# Plot the proportion of variance covered by each PC
pcvar.pct[1:10,] %>% 
  ggplot(mapping = aes(x=PC, y=pctvar)) +
  geom_col(fill="darkblue")
```

Make a tibble with the PCs and IDs
```{r}
PCs <- as_tibble(geno.pca$x) %>% # The calculated PCs
  mutate(ID=data.geno.10k$ID) %>%
  select(ID, everything())

head(PCs)
```

### PC and MDS Plots

```{r}
# Plot the PCs
PCs %>% 
  ggplot(mapping = aes(x=PC1, y=PC2)) +
  labs(title="PC1 vs PC2") +
  geom_point(color="darkblue")
  
PCs %>% 
  ggplot(mapping = aes(x=PC3, y=PC2)) +
  labs(title="PC3 vs PC2") +
  geom_point(color="darkblue")
```

```{r}
# Make an MDS Plot

# calculate the Euclidian distance between each rice variety
genDist <- as.matrix(dist(geno.numeric))

# perform multi-dimensional scaling
geno.mds <- as_tibble(cmdscale(genDist))

# Add the variety ID back into this
geno.mds$ID <- data.geno.10k$ID 
head(geno.mds) #now we have 2 dimensions + the ID

# Plot V1 vs V2 
geno.mds %>% 
  ggplot(mapping = aes(x=V1, y=V2)) +
  labs(title = "MDS Plot, V1 vs V2") +
  geom_point(color="darkblue") 
```

_The PCA and MDS plots suggest the existence of at least 3 sub. populations_


### Add Phenotypes

```{r}
data.pheno <- read_csv("../input/RiceDiversity.44K.MSU6.Phenotypes.csv")
head(data.pheno)

data.pheno.pca <- inner_join(PCs, data.pheno, by=c("ID"="NSFTVID"))
summary(data.pheno.pca)
```

```{r}
# Color PCA plot by Amylose content

data.pheno.pca %>%
  ggplot(mapping = aes(x=PC1, y=PC2, color=`Amylose content`)) +
  labs(title = "PC1 vs PC2 colored by Amylose content") +
  geom_point()

# Color PCA plot by pericarp color
data.pheno.pca %>% 
  ggplot(mapping = aes(x=PC1, y=PC2, color=`Pericarp color`)) +
  labs(title = "PC1 vs PC2 colored by Pericarp color") +  
  geom_point()

# Color PCA plot by Region
data.pheno.pca %>%
  ggplot(mapping = aes(x=PC1, y=PC2, color=`Region`)) +
  labs(title = "PC1 vs PC2 colored by Region") +
  scale_color_brewer(type="qual", palette = "Set1") +
  geom_point()
```


### Population assignments with fastStructure


##### Create the genotype file

```{r}
# Create a new Matrix to hold reformatted data
data.geno.10k.fs <- matrix("",nrow=nrow(data.geno.10k)*2,ncol=ncol(data.geno.10k)-1+6)

# for each row of genotypes, create 2 rows, one with the first allele and one with the second allele.
# also, fill the first 6 columns with the strain ID (could also be blank but we need something there)
for (i in 1:nrow(data.geno.10k)) {
  data.geno.10k.fs[(i-1)*2+1,1:6] <- data.geno.10k[[i,1]]
  data.geno.10k.fs[(i-1)*2+2,1:6] <- data.geno.10k[[i,1]]
  data.geno.10k.fs[(i-1)*2+1,-1:-6] <- substr(data.geno.10k[i,-1],1,1)
  data.geno.10k.fs[(i-1)*2+2,-1:-6] <- substr(data.geno.10k[i,-1],2,2)
}

data.geno.10k.fs[is.na(data.geno.10k.fs)] <- -9 # fastStructure's code for missing data

dim(data.geno.10k.fs)

# look at the data
data.geno.10k.fs[1:10,1:10]

write.table(data.geno.10k.fs,file="../output/rice.data.fastStructure.input.str", col.names = FALSE, row.names = FALSE, quote = FALSE)
```

##### Create the .fam file

```{r}
fam <- tibble(
  FID=data.geno.10k$ID,
  IID=data.geno.10k$ID,
  PID=0,
  MID=0,
  Sex=0,
  Ptype=-9)

head(fam)
```

```{r}
write.table(fam,file="../output/rice.data.fastStructure.input.fam",col.names = FALSE, row.names = FALSE, quote = FALSE)
```

##### Create the .bim file

```{r}
bim <- data.geno.10k.fs[,-1:-6]

colnames(bim) <- colnames(data.geno.10k)[-1]

bim[bim=="-9"] <- NA

bim <- apply(bim,2,function(x) unique(na.omit(x))) 

bim[,1:5]
```

```{r}
bim <- t(bim) %>%  # t transposes the matrix
  as_tibble() %>%
  mutate(SNP_ID=colnames(bim), cM=0) 

head(bim) # now have a table where each row is a SNP
```

```{r}
# Separate SNP_ID to get chromosome and position

bim <- bim %>% 
  separate(SNP_ID,into = c("chromosome","position"),sep="_",remove=FALSE) %>% # create a column for chromosome and position
  select(chromosome, SNP_ID, cM, position, allele1=V1, allele2=V2) # get columns in right order

head(bim)
```

```{r}
write.table(bim,file="../output/rice.data.fastStructure.input.bim",col.names = FALSE, row.names = FALSE, quote = FALSE)
```

### Running fastStructure


Running on Ubuntu v. 20.04 using WSL

The following code was executed within the fastStructure-master directory

```
python structure.py -K 4 --input=/mnt/c/Users/aangu/bioinfo/bis180L/Rice-PCA-SNPs/output/rice.data.fastStructure.input --output=/mnt/c/Users/aangu/bioinfo/bis180L/Rice-PCA-SNPs/output/rice.fastStructure.out --format=str
```
